apiserver/internal/handlers/resources/resourcemigration.go

```diff
@@ -91,120 +83,84 @@ func (h *resourcesMigrationUploadHandler) servePost(w http.ResponseWriter, r *ht
+		return empty, nil
+	}
 
-	userID := query.Get("user") // Is allowed to be blank.
+	// Get the resource UUID.
+	resUUID, err := resourceService.GetResourceUUIDByApplicationAndResourceName(
+		ctx, appName, resourceName,
+	)
+	if err != nil {
+		return empty, internalerrors.Errorf(
+			"resource upload failed: getting resource %s on application %s: %w",
+			appName, resourceName, err,
+		)
+	}
 
-	// Get the target of the upload, which is an application with or without
-	// unit.
-	target, err := getUploadTarget(ctx, applicationService, query)
+	retrievedBy, retrievedByType := determineRetrievedBy(query)
+	err = resourceService.StoreResource(ctx, resource.StoreResourceArgs{
+		ResourceUUID:    resUUID,
+		Reader:          r.Body,
+		RetrievedBy:     retrievedBy,
+		RetrievedByType: retrievedByType,
+	})
 	if err != nil {
 		return empty, internalerrors.Capture(err)
 	}
 
-	resUUID, err := resourceService.GetApplicationResourceID(ctx,
-		resource.GetApplicationResourceIDArgs{
-			ApplicationID: target.appID,
-			Name:          target.name,
-		})
+	return resourceService.GetResource(ctx, resUUID)
+}
+
+// determineRetrievedBy determines the entity that retrieved the resource using
+// the origin and user arguments on the query. If it cannot determine an entity,
+// it returns the default values.
+func determineRetrievedBy(query url.Values) (string, coreresource.RetrievedByType) {
+	rawOrigin := query.Get("origin")
+	origin, err := charmresource.ParseOrigin(rawOrigin)
 	if err != nil {
-		return empty, internalerrors.Errorf("resource upload failed: %w", err)
+		// If the origin cannot be determined, or is not present, return the
+		// default values.
+		return "", ""
 	}
 
-	if target.unitUUID != "" {
-		err := resourceService.SetUnitResource(ctx, resUUID, target.unitUUID)
-		if err != nil {
-			return empty, internalerrors.Capture(err)
-		}
-	}
-	if !isPlaceholder(query) {
-		var (
-			retrievedBy     string
-			retrievedByType coreresource.RetrievedByType
-		)
-		if target.unitUUID != "" {
-			retrievedBy = target.unitUUID.String()
-			retrievedByType = coreresource.Unit
-		} else if userID != "" {
-			retrievedBy = userID
-			retrievedByType = coreresource.User
-		} else {
-			retrievedBy = target.appID.String()
-			retrievedByType = coreresource.Application
-		}
-
-		err := resourceService.StoreResource(ctx, resource.StoreResourceArgs{
-			ResourceUUID:    resUUID,
-			Reader:          r.Body,
-			RetrievedBy:     retrievedBy,
-			RetrievedByType: retrievedByType,
-		})
-		if err != nil {
-			return empty, internalerrors.Capture(err)
+	// The user key contains the name of the entity that retrieved this
+	// resource. Confusingly, this can be an application, unit or user.
+	retrievedBy := query.Get("user")
+	switch origin {
+	case charmresource.OriginUpload:
+		return retrievedBy, coreresource.User
+	case charmresource.OriginStore:
+		// If retrievedBy has a "/" in it, it must be a unit name.
+		if strings.Contains(retrievedBy, "/") {
```

Comment:
todo: use an official unit name validator. This will be painful to track down if something ever changes.

Reply:
Good point, used `IsValidUnit` from the names package.